File: C++/Arrays/rearrangeposandneg.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
vector<int> separateNegativeAndPositive(vector<int> &nums)
{
    // Sort the vector in ascending order
    sort(nums.begin(), nums.end());

    // Return the sorted vector
    return nums;
}
vector<int> separateNegativeAndPositive(vector<int> &nums)
{
    int i = 0, j = nums.size() - 1;

    // Use two pointers to separate negative and positive integers
    while (i < j)
    {
        if (nums[i] >= 0 && nums[j] < 0)
        {
            // Swap the elements if i-th element is positive and j-th element is negative
            swap(nums[i], nums[j]);
            i++;
            j--;
        }

        // Increment i if the i-th element is already negative
        if (nums[i] < 0)
        {
            i++;
        }

        // Decrement j if the j-th element is already positive
        if (nums[j] >= 0)
        {
            j--;
        }
    }

    // Return the updated vector with negative and positive integers separated
    return nums;
}
int main()
{
    vector<int> nums = {1, -4, -2, 5, 3};
    vector<int> ans = separateNegativeAndPositive(nums);
    for (auto i : ans)
    {
        cout << i << " ";
    }
}

File: Python/Arrays/rearrangeposandneg.py
from os import *
from sys import *
from collections import *
from math import *


def separateNegativeAndPositive(nums):
    nums.sort()  # Sort the numbers in ascending order
    return nums  # Return the sorted list
    pass


def separateNegativeAndPositive(nums):
    i = 0  # Pointer for the beginning of the list
    j = len(nums) - 1  # Pointer for the end of the list

    while i < j:  # Continue until the pointers meet or cross each other
        if nums[i] >= 0 and nums[j] < 0:
            # If the number at position i is positive and the number at position j is negative,
            # swap their positions to move negative numbers to the left and positive numbers to the right
            nums[i], nums[j] = nums[j], nums[i]
            i += 1  # Move the left pointer one step forward
            j -= 1  # Move the right pointer one step backward

        if nums[i] < 0:
            i += 1  # If the number at position i is already negative, move the left pointer one step forward

        if nums[j] >= 0:
            j -= 1  # If the number at position j is already positive, move the right pointer one step backward

    return nums  # Return the modified list


nums = [1, -4, -2, 5, 3]
ans = separateNegativeAndPositive(nums)
print(ans)
